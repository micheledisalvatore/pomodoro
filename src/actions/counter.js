import {
  COUNTER_SET_NEW_SESSION,
} from '../constants/counter';

export const setNewSession = payload => ({
  type: COUNTER_SET_NEW_SESSION,
  ...payload,
});
