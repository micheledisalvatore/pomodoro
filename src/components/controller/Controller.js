import React from 'react';
import PropTypes from 'prop-types';

import { Button, PlayIcon, PauseIcon } from './Controller.style';

export const Controller = ({ isPaused, onClick }) => (
  <Button onClick={onClick}>
    { isPaused ? <PlayIcon /> : <PauseIcon /> }
  </Button>
);

Controller.propTypes = {
  isPaused: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
};

Controller.defaultProps = {
  isPaused: false,
};
