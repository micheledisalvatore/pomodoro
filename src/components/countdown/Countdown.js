import React from 'react';
import PropTypes from 'prop-types';

import {
  Container, Number, Svg, Circle,
} from './Countdown.styled';

export const Countdown = ({
  color,
  length,
  isPaused,
  timing,
}) => {
  const minutes = Math.floor(timing / 60).toString().padStart(2, '0');
  const seconds = (timing % 60).toString().padStart(2, '0');

  return (
    <Container>
      <Number color={color}>{minutes}:{seconds}</Number>
      { timing > 0 && (
        <Svg>
          <Circle duration={length} color={color} isPaused={isPaused} />
        </Svg>
      )}
    </Container>
  );
};

Countdown.propTypes = {
  color: PropTypes.string.isRequired,
  isPaused: PropTypes.bool.isRequired,
  length: PropTypes.number.isRequired,
  timing: PropTypes.number.isRequired,
};

export const mapStateToProps = ({
  session: {
    color,
    isPaused,
    length,
    timing,
  },
}) => ({
  color,
  length,
  isPaused,
  timing,
});
