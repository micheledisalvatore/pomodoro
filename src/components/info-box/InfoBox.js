import React from 'react';
import PropTypes from 'prop-types';

import { Container, Current } from './InfoBox.styled';

export const InfoBox = ({ current, total, title }) => (
  <Container>
    <h3>{title}</h3>
    <Current>{current}</Current><span>/{total}</span>
  </Container>
);

InfoBox.propTypes = {
  title: PropTypes.string.isRequired,
  current: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
};
