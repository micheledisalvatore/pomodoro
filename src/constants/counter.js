export const COUNTER_SET_NEW_SESSION = 'COUNTER::SET_NEW_SESSION';

export const COUNTER_SESSIONS = 4;
export const COUNTER_GOALS = 12;
