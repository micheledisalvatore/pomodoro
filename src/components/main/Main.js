import React from 'react';
import PropTypes from 'prop-types';

import { COUNTER_GOALS, COUNTER_SESSIONS } from '../../constants/counter';

import Countdown from '../countdown';
import Controller from '../controller';
import InfoBox from '../info-box';

import { Container } from './Main.styled';

export const Main = ({ goalNumber, sessionNumber }) => (
  <Container>
    <Countdown />
    <Controller />
    <div>
      <InfoBox current={sessionNumber} total={COUNTER_SESSIONS} title="Round" />
      <InfoBox current={goalNumber} total={COUNTER_GOALS} title="Goal" />
    </div>
  </Container>
);

Main.propTypes = {
  goalNumber: PropTypes.number.isRequired,
  sessionNumber: PropTypes.number.isRequired,
};

export const mapStateToProps = ({
  counter: {
    goalNumber,
    sessionNumber,
  },
}) => ({
  goalNumber,
  sessionNumber,
});
