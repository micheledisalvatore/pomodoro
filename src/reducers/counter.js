import { handleActions } from 'redux-actions';

import { SESSIONS } from '../constants/session';
import { COUNTER_SET_NEW_SESSION } from '../constants/counter';

const defaultState = {
  sessionTypeLabel: SESSIONS.FOCUS.label,
  sessionNumber: 0,
  goalNumber: 0,
};

const setNewSession = (state, { type, ...payload }) => ({
  ...state,
  ...payload,
});

export default handleActions({
  [COUNTER_SET_NEW_SESSION]: setNewSession,
}, defaultState);
