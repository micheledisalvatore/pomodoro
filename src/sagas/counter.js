import {
  fork,
  put,
  select,
  takeLatest,
} from 'redux-saga/effects';

import { COUNTER_SESSIONS, COUNTER_GOALS } from '../constants/counter';
import { setNewSession } from '../actions/counter';
import { SESSION_COMPLETE, SESSIONS } from '../constants/session';
import { startSession } from '../actions/session';

const getCounterState = ({ counter }) => counter;

function* start() {
  const {
    goalNumber: prevGoalNumber,
    sessionNumber: prevSessionNumber,
    sessionTypeLabel: prevSessionTypeLabel,
  } = yield select(getCounterState);

  let goalNumber = prevGoalNumber;
  let sessionNumber = prevSessionNumber;

  const isLongBreakSession = prevSessionTypeLabel === SESSIONS.LONG_BREAK.label;
  const shouldResetCounter = isLongBreakSession && goalNumber === COUNTER_GOALS;
  let sessionType;

  switch (prevSessionTypeLabel) {
    case SESSIONS.FOCUS.label:
      if (sessionNumber < COUNTER_SESSIONS) {
        sessionType = SESSIONS.SHORT_BREAK;
      } else {
        sessionType = SESSIONS.LONG_BREAK;
      }
      break;
    case SESSIONS.LONG_BREAK.label:
      sessionType = SESSIONS.FOCUS;
      sessionNumber = 1;
      goalNumber = shouldResetCounter ? 1 : goalNumber;
      break;
    default:
      sessionType = SESSIONS.FOCUS;
      sessionNumber += 1;
      goalNumber += 1;
      break;
  }

  yield put(setNewSession({
    goalNumber,
    sessionNumber,
    sessionTypeLabel: sessionType.label,
  }));

  if (!shouldResetCounter) {
    yield put(startSession(sessionType));
  }
}

function* watchActions() {
  yield takeLatest(SESSION_COMPLETE, start);
}

export default fork(watchActions);
