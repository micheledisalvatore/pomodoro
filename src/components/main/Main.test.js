/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import { COUNTER_GOALS, COUNTER_SESSIONS } from '../../constants/counter';

import Countdown from '../countdown';
import Controller from '../controller';
import InfoBox from '../info-box';

import { Main, mapStateToProps } from './Main';

describe('Given a Main component', () => {
  let component;
  let wrapper;

  const requiredProps = {
    sessionNumber: 1,
    goalNumber: 2,
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <Main {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should show the Countdown component', () => {
      expect(wrapper.find(Countdown)).toExist();
    });

    it('should show the Controller component', () => {
      expect(wrapper.find(Controller)).toExist();
    });

    it('should show the first InfoBox component with Round title', () => {
      expect(wrapper.find(InfoBox).at(0)).toHaveProp('title', 'Round');
    });

    it('should show the first InfoBox component with current prop set to 1', () => {
      expect(wrapper.find(InfoBox).at(0)).toHaveProp('current', 1);
    });

    it('should show the first InfoBox component with total prop set to the constant value of COUNTER_SESSIONS', () => {
      expect(wrapper.find(InfoBox).at(0)).toHaveProp('total', COUNTER_SESSIONS);
    });

    it('should show the first InfoBox component with Goal title', () => {
      expect(wrapper.find(InfoBox).at(1)).toHaveProp('title', 'Goal');
    });

    it('should show the first InfoBox component with current prop set to 1', () => {
      expect(wrapper.find(InfoBox).at(1)).toHaveProp('current', 2);
    });

    it('should show the first InfoBox component with total prop set to the constant value of COUNTER_GOALS', () => {
      expect(wrapper.find(InfoBox).at(1)).toHaveProp('total', COUNTER_GOALS);
    });
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        counter: {
          goalNumber: 'foo',
          sessionNumber: 'bar',
        },
      })).toEqual({
        goalNumber: 'foo',
        sessionNumber: 'bar',
      });
    });
  });
});
