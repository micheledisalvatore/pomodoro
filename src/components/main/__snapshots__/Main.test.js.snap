// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Main component when it is rendered should match the snapshot 1`] = `
<styled.div>
  <Connect(Countdown) />
  <Connect(Controller) />
  <div>
    <InfoBox
      current={1}
      title="Round"
      total={4}
    />
    <InfoBox
      current={2}
      title="Goal"
      total={12}
    />
  </div>
</styled.div>
`;
